benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,228 @@
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return path
+
+
+class PersistenceHandlingInterface(Protocol):
+    """Interface for interacting with storage."""
+
+    def write_object(self, path_constructor: PathConstructor, object: dict) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        If the respective object exists, it will be overwritten.
+
+        Args:
+            path_constructor: The PathConstructor to create the path for the object.
+            object: The object to be persisted.
+        """
+
+
+@define
+class S3PersistenceHandler:
+    """Class for persisting objects in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session: Session = boto3.session.Session()
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        """Get the bucket name from the environment variables."""
+        if "BAYBE_PERFORMANCE_PERSISTENCE_PATH" not in os.environ:
+            raise ValueError(
+                "No S3 bucket name provided. Please provide the "
+                + "bucket name by setting the environment variable "
+                + "BAYBE_PERFORMANCE_PERSISTENCE_PATH."
+            )
+        return os.environ["BAYBE_PERFORMANCE_PERSISTENCE_PATH"]
+
+    def write_object(self, path_constructor: PathConstructor, object: dict) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        This method will store an JSON serializable dict in an S3 bucket.
+        The S3-key of the Java Script Notation Object will be created by
+        the path object. I the key already exists, it will be overwritten.
+
+        Args:
+            path_constructor: The PathConstructor to create the path for the object.
+            object: The object to be persisted.
+        """
+        client = self._object_session.client("s3")
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=path_constructor.get_path(strategy=PathStrategy.HIERARCHICAL),
+            Body=json.dumps(object),
+            ContentType="application/json",
+        )
+
+
+@define
+class LocalFileSystemPersistenceHandler:
+    """Class for persisting JSON serializable dicts locally."""
+
+    folder_path_prefix: str = field(validator=instance_of(str), default="./")
```

Comment:
I guess this will also break on Windows. Can you use a `pathlib` solution?

Reply:
I use Pathlib in the function body but now have changed the Prefix from string to pathlib.

Reply:
👍🏼 have further changed the validator into a converter, which makes it more flexible